File: src/degenbot/cli/pool.py
"""CLI commands for pool state queries.

Task `JJ232N` (epic `2SFL6I`): `pool_update` is now a thin boot + hand-off
to the Rust-owned chunk loop (`degenbot_rs.run_pool_update`, Task
`QZHNZQ`). Python is a driver shell -- config bootstrap, SIGINT -> cancel
flag, tqdm-on-callback, + a user-facing summary. The chunk loop, the RPC
fetches, the decode, the DB writes, + the per-chunk transaction all live in
the Rust core (`degenbot-pool-updater`, Task `CKXCOB`). The SQLAlchemy
session-for-writes, the per-call `db_*` dispatch, the
`fresh_last_update_block` re-read workaround, + the per-event Python tqdm
iteration are all retired (migration-guide `pool-updater-chunk-atomicity`
section 3.1 + section 4 Task 5).

The standalone PyO3 seams (`db_apply_v*_liquidity_updates`,
`db_upsert_v*_pools`, etc.) stay in `degenbot_rs` for ad-hoc/test uses
(`tests/rust/test_discovery_seam.py` exercises the `update_v*_pools`
shells in `pool_updater_configs.py`); only the chunk-loop *usages* in this
file are gone.
"""

from __future__ import annotations

import signal
from typing import TYPE_CHECKING, Any, Literal, cast

import click
import tqdm
from tqdm.contrib.logging import logging_redirect_tqdm

from degenbot.cli import cli
from degenbot.cli.utils import get_provider_from_config
from degenbot.config import resolve_http_rpc_uri
from degenbot.degenbot_rs import (
    CancelHandle,
    run_pool_update,
    verify_v3_liquidity_map,
    verify_v4_liquidity_map,
)
from degenbot.logging import logger
from degenbot.provider.block_helpers import get_number_for_block_identifier

if TYPE_CHECKING:
    from degenbot.bot import Bot

# Block tags the `--to-block` option accepts (mirrors the prior Python loop).
# A pure tag (no offset) resolves to `None` -> the Rust core fetches the chain
# tip via `eth_blockNumber` itself (no Python-side RPC round-trip). A tag with
# an offset (`latest:-64`, `safe:128`) is resolved in Python to a concrete
# block number (the Rust core takes `Option<u64>`, not a tag string).
_BlockTag = Literal["latest", "earliest", "pending", "safe", "finalized"]

_BLOCK_TAGS: frozenset[_BlockTag] = frozenset(
    {"latest", "earliest", "pending", "safe", "finalized"},
)


@cli.group()
def pool() -> None:
    """Pool commands."""


@pool.command("update")
@click.option(
    "--chunk",
    "chunk_size",
    default=10_000,
    show_default=True,
    help="The maximum number of blocks to process before committing changes to the database.",
)
@click.option(
    "--to-block",
    "to_block",
    default="latest:-64",
    show_default=True,
    help=(
        "The last block in the update range. Must be a valid block identifier: "
        "'earliest', 'finalized', 'safe', 'latest', 'pending'. An identifier can be given with an "
        "optional offset, e.g. 'latest:-64' stops 64 blocks before the chain tip, "
        "'safe:128' stops 128 blocks after the last 'safe' block."
    ),
)
@click.option(
    "--verify/--no-verify",
    "verify",
    default=False,
    show_default=True,
    help=(
        "Run the pre-commit on-chain-truth gate (Full per-pool per-chunk "
        "verification) before each chunk's liquidity persist commits. A "
        "divergence rolls back the chunk + does NOT advance "
        "last_update_block (the corrupt state never lands). Catches the "
        "6SRJRL ghost-row class + other apply-math corruption before write."
    ),
)
@click.pass_obj
def pool_update(bot: Bot, chunk_size: int, to_block: str, verify: bool) -> None:  # noqa: FBT001
    """Update liquidity pool information for activated exchanges.

    Boot + hand-off: read the bot config, install a SIGINT -> cancel-flag
    handler, build a tqdm-ticking progress callback, + delegate the whole
    chunk loop to the Rust core (`degenbot_rs.run_pool_update`). The core
    owns the RPC fetches, the decode, the per-chunk transaction (atomicity),
    + the `last_update_block` stamp (restart-invariance). The GIL is
    released across the whole run; only the progress callback re-acquires
    it briefly, once per chunk.

    Raises:
        ValueError: For a malformed `--to-block` or an RPC/DB failure (the
            in-flight chunk is rolled back before returning; committed
            chunks stay durable).

    """
    chain_id = bot.config.default_chain_id
    if chain_id is None:
        msg = (
            "Bot requires a default_chain_id in the config. Set "
            "`default_chain_id` in your config file or pass a config with it set."
        )
        raise ValueError(msg)

    database_path = str(bot.config.database.path)
    rpc_url = resolve_http_rpc_uri(chain_id, config=bot.config)
    resolved_to_block = _resolve_to_block(to_block, chain_id=chain_id, bot=bot)

    handle = CancelHandle()
    prior_int_handler = signal.getsignal(signal.SIGINT)
    n_chunks = 0

    def _on_sigint(*_args: Any) -> None:
        # Cooperative cancel: the Rust loop polls the flag between chunks (NOT
        # mid-chunk) so a SIGINT never breaks chunk atomicity -- the in-flight
        # chunk completes (commit OR rollback) before the run returns
        # (migration-guide section 3.3 interrupt contract).
        handle.cancel()

    def _on_progress(progress: dict[str, Any]) -> None:
        # tqdm ticks once per chunk (not per event) -- the Rust core reports the
        # chunk boundary; this closure updates the bar's position + postfix.
        nonlocal n_chunks
        n_chunks += 1
        if not progress["committed"]:
            # A rolled-back chunk: don't advance the bar (the next run
            # re-processes it); show the skip in the postfix so the user sees it.
            pbar.set_postfix_str(
                f"chunk {progress['chunk_start']}-{progress['chunk_end']} rolled back",
                refresh=True,
            )
            return
        delta = progress["chunk_end"] - progress["chunk_start"] + 1
        pbar.update(delta)
        pbar.set_postfix_str(
            f"+{progress['pools_written']} pools, "
            f"{progress['liquidity_apply_count']} liq applies "
            f"(chunk {n_chunks})",
            refresh=True,
        )

    total = None  # indeterminate: the core resolves the tip; shows a per-chunk
    # postfix (pools, liquidity applies, chunk number), not a %.
    pbar = tqdm.tqdm(
        desc="Processing new blocks",
        total=total,
        bar_format="{desc}: {n_fmt} blocks |{bar}| {postfix}",
        leave=False,
    )

    signal.signal(signal.SIGINT, _on_sigint)
    try:
        with logging_redirect_tqdm(loggers=[logger]):
            report = run_pool_update(
                database_path=database_path,
                chain_id=chain_id,
                to_block=resolved_to_block,
                chunk_size=chunk_size,
                rpc_url=rpc_url,
                progress_callback=_on_progress,
                cancel_handle=handle,
                verify=verify,
            )
    finally:
        # Restore the prior SIGINT handler (or a default- disposition if there
        # wasn't one) so a subsequent Ctrl+C in the same shell behaves normally.
        signal.signal(signal.SIGINT, prior_int_handler)
        pbar.close()

    click.echo(
        f"Chain {report['chain_id']}: advanced {report['from_block']}->"
        f"{report['to_block']} in {report['chunks_committed']} chunks "
        f"({report['total_pools_written']} pools written, "
        f"{report['total_liquidity_applies']} liquidity applies).",
    )


def _resolve_to_block(to_block: str, *, chain_id: int, bot: Bot) -> int | None:
    """Resolve the `--to-block` CLI string to `int | None`.

    `None` means "advance to the chain tip" -- the Rust core fetches the tip
    via `eth_blockNumber` itself (no Python-side round-trip). A pure block tag
    (`latest`, `safe`, ...) with no offset maps to `None`; a tag with an offset
    (`latest:-64`, `safe:128`) or a concrete integer resolves to a specific
    block number via the provider.

    Returns:
        `int` for a concrete block number; `None` to let the Rust core fetch
        the chain tip.

    Raises:
        ValueError: For a malformed tag.

    """
    if to_block.isdigit():
        return int(to_block)

    if ":" in to_block:
        parts = to_block.split(":", 1)
        block_tag, offset = parts[0], parts[1]
        block_offset = int(offset.strip())
    else:
        block_tag = to_block
        block_offset = 0

    if block_tag not in _BLOCK_TAGS:
        msg = f"Invalid block tag: {block_tag}"
        raise ValueError(msg)

    if block_offset == 0:
        # Pure tag -> let the Rust core resolve the tip.
        return None

    # Tag + offset -> resolve to a concrete block number in Python (the core
    # takes `Option<u64>`, not a tag string).
    provider = get_provider_from_config(chain_id=chain_id, config=bot.config)
    resolved = get_number_for_block_identifier(
        identifier=cast("_BlockTag", block_tag),
        provider=provider,
    )
    return int(resolved) + block_offset


@pool.command("verify")
@click.option(
    "--rpc-url",
    "rpc_url",
    required=True,
    help="The HTTP RPC endpoint to read on-chain truth from.",
)
@click.option(
    "--chain",
    "chain_id",
    type=int,
    required=True,
    help="The chain id the pool lives on.",
)
@click.option(
    "--block",
    "block_number",
    required=True,
    type=int,
    help="The block number to read on-chain truth at.",
)
@click.option(
    "--pool",
    "pool",
    required=True,
    help=(
        "The pool to verify. V3: the pool contract address. "
        "V4: the PoolId (pool_hash, bytes32 hex 0x…)."
    ),
)
@click.option(
    "--family",
    "family",
    type=click.Choice(["v3", "v4"]),
    required=True,
    help="The pool family (selects ticks()/tickBitmap() vs PoolManager extsload).",
)
@click.option(
    "--pool-manager",
    "pool_manager",
    default=None,
    help=(
        "(V4 only) The deployed V4 PoolManager singleton address (the V4 "
        "exchange's factory). Required for --family v4."
    ),
)
@click.pass_obj
def pool_verify(  # noqa: PLR0917
    bot: Bot,
    rpc_url: str,
    chain_id: int,
    block_number: int,
    pool: str,
    family: str,
    pool_manager: str | None,
) -> None:
    """Verify a pool's committed liquidity map against on-chain truth.

    Stand-alone read-only check (no event apply, no SQL writes): opens the DB,
    fetches the pool's committed tick + bitmap rows, and compares the FULL
    map against on-chain truth (V3: ``ticks()``/``tickBitmap()``; V4:
    ``PoolManager.extsload``) at ``--block``. Prints GREEN if the DB matches
    the chain, else the named divergence list (bisect-able triage).

    This is the ad-hoc / spot-check sibling of the pre-commit gate
    (``pool update --verify``); the gate runs the SAME compare before the
    write commits, while this reads the already-committed state.

    Raises:
        ValueError: For a DB or RPC failure, or a missing V4 pool-manager.

    """
    database_path = str(bot.config.database.path)
    if family == "v3":
        divergences = verify_v3_liquidity_map(
            database_path=database_path,
            rpc_url=rpc_url,
            chain_id=chain_id,
            pool_address=pool,
            block_number=block_number,
        )
    else:
        if pool_manager is None:
            msg = "--pool-manager is required for --family v4 (the PoolManager singleton)."
            raise ValueError(msg)
        divergences = verify_v4_liquidity_map(
            database_path=database_path,
            rpc_url=rpc_url,
            chain_id=chain_id,
            pool_hash=pool,
            pool_manager_address=pool_manager,
            block_number=block_number,
        )
    if not divergences:
        click.echo(f"GREEN: {family} pool {pool} matches on-chain truth at block {block_number}.")
        return
    click.echo(
        f"RED: {len(divergences)} divergence(s) for {family} pool {pool} at block {block_number}:"
    )
    for d in divergences:
        variant = d["variant"]
        if variant in {"TickGross", "TickNet"}:
            click.echo(
                f"  tick {d['tick']}: {variant} expected={d['expected']} actual={d['actual']}"
            )
        elif variant == "BitmapWord":
            click.echo(
                f"  word {d['word']}: {variant} expected={d['expected']} actual={d['actual']}"
            )
        elif variant == "TickCallReverted":
            click.echo(f"  tick {d['tick']}: {variant}")
        elif variant == "BitmapCallReverted":
            click.echo(f"  word {d['word']}: {variant}")

File: rust/crates/degenbot-cli-core/src/pool.rs
//! The `pool` command arms (ADR-051 D1;).
//!
//! Ports `cli/pool.py` arm for arm:
//!
//! - `pool update` — the chunk-loop hand-off to
//!   [`run_pool_update`](degenbot_pool_updater::run_pool_update), with the
//!   `--chunk` / `--to-block` / `--verify-chunk` / `--verify-all` /
//!   `--verify-all-interval` flags 1:1. Progress stays a no-op sink here
//!   (ADR-051 D9: the bin paints it); a cooperative cancel returns a friendly
//!   cancelled report (exit 0), matching the Python `RuntimeError` guard.
//! - `pool verify` — the read-only, ad-hoc sibling: fetch the COMMITTED
//!   liquidity map, compare against on-chain truth at `--block`, and render
//!   GREEN / the named divergence list.

use std::str::FromStr;
use std::sync::Arc;

use alloy::primitives::{Address, B256};
use degenbot_db::{ComputedLiquidityUpdate, DegenbotDb};
use degenbot_pool_updater::{
    run_pool_update, verify_v3_liquidity_map_on_chain, verify_v4_liquidity_map_on_chain,
    NoProgress, RunError,
};

use crate::block::{parse_to_block, resolve_to_block};
use crate::cancel::CancelHandle;
use crate::context::CliContext;
use crate::error::CliError;
use crate::prompt::{PromptPlan, Prompter};
use crate::report::PoolReport;

/// The RPC retry budget the verify arm's spot reads use.
const RPC_MAX_RETRIES: u32 = 5;

/// The pool family `--family` selects.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PoolFamily {
    /// V3 (`ticks()`/`tickBitmap()`).
    V3,
    /// V4 (`PoolManager` `extsload`).
    V4,
}

impl PoolFamily {
    /// The wire spelling (`v3` / `v4`).
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::V3 => "v3",
            Self::V4 => "v4",
        }
    }
}

/// The `pool` command group.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PoolCommand {
    /// Advance every active exchange's liquidity state to `to_block`.
    Update {
        /// Max blocks per chunk before committing.
        chunk_size: u64,
        /// The raw `--to-block` identifier (tag / tag:offset / integer).
        to_block: String,
        /// Run the pre-commit per-chunk on-chain-truth gate.
        verify_chunk: bool,
        /// Run the pre-commit market-wide verification at the interval + completion.
        verify_all: bool,
        /// The block interval for the `--verify-all` gate.
        verify_all_interval: u64,
    },
    /// Verify a committed pool's liquidity map against on-chain truth.
    Verify {
        /// The HTTP RPC endpoint.
        rpc_url: String,
        /// The chain the pool lives on.
        chain_id: i64,
        /// The block number to read on-chain truth at.
        block_number: u64,
        /// V3 pool address, or V4 `PoolId` (bytes32 hex).
        pool: String,
        /// The pool family.
        family: PoolFamily,
        /// (V4 only) the `PoolManager` singleton.
        pool_manager: Option<String>,
    },
}

impl PoolCommand {
    /// The arm's confirmation policy: neither pool arm prompts.
    #[must_use]
    pub const fn prompt_plan(&self, _ctx: &CliContext<'_>) -> PromptPlan {
        PromptPlan::None
    }
}

/// The resolved verify target for a family.
enum VerifyTarget {
    /// A V3 pool contract address.
    V3(Address),
    /// A V4 `PoolManager` + `PoolId`.
    V4 { manager: Address, pool_id: B256 },
}

/// Execute a `pool` command.
///
/// # Errors
///
/// [`CliError::InvalidBlockTag`] / [`CliError::InvalidArgument`] /
/// [`CliError::InvalidAddress`] for malformed inputs, [`CliError::Config`] for
/// an unresolved driver-domain value, and [`CliError::PoolUpdate`] for a core
/// failure.
pub(crate) fn execute(
    command: &PoolCommand,
    ctx: &CliContext<'_>,
    _prompter: &dyn Prompter,
    cancel: &CancelHandle,
) -> Result<PoolReport, CliError> {
    match command {
        PoolCommand::Update {
            chunk_size,
            to_block,
            verify_chunk,
            verify_all,
            verify_all_interval,
        } => update(
            ctx,
            cancel,
            *chunk_size,
            to_block,
            *verify_chunk,
            *verify_all,
            *verify_all_interval,
        ),
        PoolCommand::Verify {
            rpc_url,
            chain_id,
            block_number,
            pool,
            family,
            pool_manager,
        } => verify(
            ctx,
            rpc_url,
            *chain_id,
            *block_number,
            pool,
            *family,
            pool_manager.as_deref(),
        ),
    }
}

/// `pool update`.
fn update(
    ctx: &CliContext<'_>,
    cancel: &CancelHandle,
    chunk_size: u64,
    to_block: &str,
    verify_chunk: bool,
    verify_all: bool,
    verify_all_interval: u64,
) -> Result<PoolReport, CliError> {
    let database_path = ctx.database_path().value;
    let chain_id = ctx.chain_id()?.value;
    let rpc_url = ctx.node_http_uri()?.value;
    // Self-serve registration: every supported exchange pair not found in the
    // DB registers inactive, so the update never depends on prior CREATEs.
    crate::registrations::ensure_supported_registrations(&database_path)?;
    let resolved = resolve_to_block(parse_to_block(to_block)?, &rpc_url)?;
    let chain = i64::try_from(chain_id)
        .map_err(|_| CliError::InvalidArgument(format!("chain id {chain_id} is out of range")))?;
    let interval = if verify_all {
        Some(verify_all_interval)
    } else {
        None
    };
    match run_pool_update(
        &database_path,
        chain,
        resolved,
        chunk_size,
        &rpc_url,
        cancel.flag(),
        Arc::new(NoProgress),
        verify_chunk,
        interval,
        verify_all,
    ) {
        Ok(report) => Ok(PoolReport::Updated {
            chain_id: report.chain_id,
            from_block: report.from_block,
            to_block: report.to_block,
            chunks_committed: report.chunks_committed,
            total_pools_written: report.total_pools_written,
            total_liquidity_applies: report.total_liquidity_applies,
        }),
        Err(RunError::Cancelled) => Ok(PoolReport::UpdateCancelled { chain_id: chain }),
        Err(err) => Err(CliError::PoolUpdate(err)),
    }
}

/// `pool verify`.
fn verify(
    ctx: &CliContext<'_>,
    rpc_url: &str,
    chain_id: i64,
    block_number: u64,
    pool: &str,
    family: PoolFamily,
    pool_manager: Option<&str>,
) -> Result<PoolReport, CliError> {
    let database_path = ctx.database_path().value;
    let (computed, target) = {
        let (db, _state) = DegenbotDb::open(&database_path)?;
        let conn = db.lock();
        fetch_verify_state(&conn, chain_id, pool, family, pool_manager)?
    };
    let divergences = match target {
        VerifyTarget::V3(address) => crate::block::block_on(async {
            let provider = degenbot_rpc::provider::AlloyProvider::new(rpc_url, RPC_MAX_RETRIES)
                .await
                .map_err(|err| CliError::BlockResolution(err.to_string()))?;
            verify_v3_liquidity_map_on_chain(&provider, address, &computed, block_number)
                .await
                .map_err(CliError::PoolUpdate)
        })??,
        VerifyTarget::V4 { manager, pool_id } => crate::block::block_on(async {
            let provider = degenbot_rpc::provider::AlloyProvider::new(rpc_url, RPC_MAX_RETRIES)
                .await
                .map_err(|err| CliError::BlockResolution(err.to_string()))?;
            verify_v4_liquidity_map_on_chain(&provider, manager, pool_id, &computed, block_number)
                .await
                .map_err(CliError::PoolUpdate)
        })??,
    };
    Ok(PoolReport::Verified {
        pool: pool.to_string(),
        family,
        block_number,
        divergences,
    })
}

/// Fetch the pool's committed liquidity map + resolve the verify target.
fn fetch_verify_state(
    conn: &rusqlite::Connection,
    chain_id: i64,
    pool: &str,
    family: PoolFamily,
    pool_manager: Option<&str>,
) -> Result<(ComputedLiquidityUpdate, VerifyTarget), CliError> {
    match family {
        PoolFamily::V3 => {
            let address: Address = pool
                .parse()
                .map_err(|_| CliError::InvalidAddress(pool.to_string()))?;
            let key = address.to_checksum(None);
            let state = DegenbotDb::fetch_v3_pool_update_state_on_conn(conn, chain_id, &key)?
                .ok_or_else(|| {
                    CliError::InvalidArgument(format!(
                        "v3 pool {key} not found on chain {chain_id}"
                    ))
                })?;
            let (tick_bitmap, tick_data) =
                DegenbotDb::fetch_v3_liquidity_map_on_conn(conn, state.pool_id)?;
            Ok((
                ComputedLiquidityUpdate {
                    pool_id: state.pool_id,
                    tick_spacing: state.tick_spacing,
                    tick_data,
                    tick_bitmap,
                    last_event: None,
                },
                VerifyTarget::V3(address),
            ))
        }
        PoolFamily::V4 => {
            let manager = pool_manager.ok_or_else(|| {
                CliError::InvalidArgument(
                    "--pool-manager is required for --family v4 (the PoolManager singleton)."
                        .to_string(),
                )
            })?;
            let manager: Address = manager
                .parse()
                .map_err(|_| CliError::InvalidAddress(manager.to_string()))?;
            let pool_id = B256::from_str(pool.strip_prefix("0x").unwrap_or(pool))
                .map_err(|_| CliError::InvalidAddress(pool.to_string()))?;
            let state = DegenbotDb::fetch_v4_pool_update_state_on_conn(conn, pool, chain_id)?
                .ok_or_else(|| {
                    CliError::InvalidArgument(format!(
                        "v4 pool {pool} not found on chain {chain_id}"
                    ))
                })?;
            let (tick_bitmap, tick_data) =
                DegenbotDb::fetch_v4_liquidity_map_on_conn(conn, state.pool_id)?;
            Ok((
                ComputedLiquidityUpdate {
                    pool_id: state.pool_id,
                    tick_spacing: state.tick_spacing,
                    tick_data,
                    tick_bitmap,
                    last_event: None,
                },
                VerifyTarget::V4 { manager, pool_id },
            ))
        }
    }
}
